database/pebble/iterator.go

```diff
@@ -0,0 +1,159 @@
+)
+
+var (
+	_ database.Iterator = (*iter)(nil)
+
+	errCouldntGetValue = errors.New("couldnt get iterator value")
+	errNoPointKey      = errors.New("iterator has no point key")
+)
+
+type iter struct {
+	// [lock] ensures that only one goroutine can access [iter] at a time.
+	// Note that [Database.Close] calls [iter.Release] so we need [lock] to ensure
+	// that the user and [Database.Close] don't execute [iter.Release] concurrently.
+	// Invariant: [Database.lock] is never grabbed while holding [lock].
+	lock sync.Mutex
+
+	db   *Database
+	iter *pebble.Iterator
+
+	initialized bool
+	closed      bool
+	err         error
+
+	hasNext bool
+	nextKey []byte
+	nextVal []byte
+}
+
+// Must not be called with [db.lock] held.
+func (it *iter) Next() bool {
+	it.db.lock.RLock()
+	dbClosed := it.db.closed
+	it.db.lock.RUnlock()
+
+	it.lock.Lock()
+	defer it.lock.Unlock()
+
+	switch {
+	case it.closed || dbClosed:
+		it.hasNext = false
+		it.err = database.ErrClosed
+		return false
+	case it.err != nil:
+		it.hasNext = false
+		return false
+	case !it.initialized:
+		it.hasNext = it.iter.First()
+		it.initialized = true
+	default:
+		it.hasNext = it.iter.Next()
+	}
+
+	if !it.hasNext {
+		return false
+	}
+
+	// Set the next key.
+	it.nextKey = it.iter.Key()
+
+	// Set the next value.
+	// TODO is the below necessary?
+	// Need to make sure the following invariant for [i.iter.ValueAndErr] holds:
+	// "REQUIRES: iter.Error() == nil and HasPointAndRange() returns true for hasPoint."
+	hasPoint, _ := it.iter.HasPointAndRange()
+	if !hasPoint {
+		it.hasNext = false
+		it.err = fmt.Errorf("%w: %w", errCouldntGetValue, errNoPointKey)
+		return false
+	}
+
+	if err := it.iter.Error(); err != nil {
+		it.hasNext = false
+		it.err = fmt.Errorf("%w: %w", errCouldntGetValue, err)
+		return false
+	}
+
+	var err error
+	it.nextVal, err = it.iter.ValueAndErr()
+	if err != nil {
+		it.hasNext = false
+		it.err = fmt.Errorf("%w: %w", errCouldntGetValue, err)
+		return false
+	}
+
+	return true
+}
+
+func (it *iter) Error() error {
+	it.lock.Lock()
+	defer it.lock.Unlock()
+
+	if it.err != nil {
+		return it.err
+	}
+	if it.closed {
+		return nil
+	}
+	return updateError(it.iter.Error())
+}
+
+func (it *iter) Key() []byte {
+	it.lock.Lock()
+	defer it.lock.Unlock()
+
+	if !it.hasNext {
+		return nil
+	}
+	return slices.Clone(it.nextKey)
+}
+
+func (it *iter) Value() []byte {
+	it.lock.Lock()
+	defer it.lock.Unlock()
+
+	if !it.hasNext {
+		return nil
+	}
+	return slices.Clone(it.nextVal)
+}
+
+func (it *iter) Release() {
+	it.db.lock.Lock()
+	defer it.db.lock.Unlock()
+
+	it.lock.Lock()
+	defer it.lock.Unlock()
+
+	it.release()
+}
+
+// Assumes [it.lock] and [it.db.lock] are held.
+func (it *iter) release() {
+	if it.closed {
+		return
+	}
+
+	// Remove the iterator from the list of open iterators.
+	it.db.openIterators.Remove(it)
+
+	it.closed = true
+	if err := it.iter.Close(); err != nil {
+		it.err = updateError(err)
+	}
```

Comment:
What is the behavior of `Key` and `Value` after `Release` is called? We have a test for this similar flow: `TestIteratorErrorAfterRelease`... Seems like we should document the expected behavior (or document that it is unexpected to call `Key` + `Value` after `Release`.

Reply:
Added `// Behavior is underined after Release is called.` to `Key()` and `Value()` on `database.Iterator`